portal/flatpak-portal.c

```diff
@@ -247,10 +251,218 @@ typedef struct
+
+  signal_variant = g_variant_ref_sink (g_variant_new ("(uu)", pid, relative_child_pid));
+  g_dbus_connection_emit_signal (session_bus,
+                                 pid_data->client,
+                                 "/org/freedesktop/portal/Flatpak",
+                                 "org.freedesktop.portal.Flatpak",
+                                 "SpawnStarted",
+                                 signal_variant,
+                                 NULL);
+}
+
+static void
+bwrapinfo_changed (GFileMonitor         *monitor,
+                   GFile                *file,
+                   GFile                *other_file,
+                   GFileMonitorEvent     event_type,
+                   BwrapinfoMonitorData *data)
+{
+  g_autofree char *path = NULL;
+
+  switch (event_type)
+    {
+    case G_FILE_MONITOR_EVENT_DELETED:
+      path = g_file_get_path (file);
+      g_error ("%s deleted before child pid was added", path);
+      g_file_monitor_cancel (monitor);
+      break;
+    case G_FILE_MONITOR_EVENT_CHANGES_DONE_HINT:
+      check_child_pid_status (monitor, data);
+      break;
+    default:
+      break;
+    }
+}
+
+static void
+instance_id_read_finish (GObject      *source,
+                         GAsyncResult *res,
+                         gpointer      user_data)
+{
+  g_autoptr(GInputStream) stream = NULL;
+  g_autofree InstanceIdReadData *data = NULL;
+  g_autofree char *buffer = NULL;
+  g_autoptr(FlatpakInstance) instance = NULL;
+  g_autoptr(GFile) instance_dir = NULL;
+  g_autoptr(GFile) bwrapinfo = NULL;
+  g_autoptr(GError) error = NULL;
+  GFileMonitor *monitor = NULL;
+  BwrapinfoMonitorData *monitor_data = NULL;
+  gssize bytes_read;
+
+  stream = G_INPUT_STREAM (source);
+  data = (InstanceIdReadData *) user_data;
+  buffer = g_steal_pointer (&data->buffer);
+
+  bytes_read = g_input_stream_read_finish (stream, res, &error);
+  if (bytes_read <= 0)
+    {
+      /* 0 means EOF, so the process could never have been started. */
+      if (bytes_read == -1)
+        g_error ("Failed to read instance id: %s", error->message);
+
+      return;
+    }
+
+  buffer[bytes_read] = 0;
+
+  instance = flatpak_instance_new_for_id (buffer);
+  instance_dir = g_file_new_for_path (flatpak_instance_get_dir (instance));
+  bwrapinfo = g_file_get_child (instance_dir, "bwrapinfo.json");
+
+  g_warn_if_fail (g_file_query_exists (bwrapinfo, NULL));
+
+  monitor = g_file_monitor_file (bwrapinfo, G_FILE_MONITOR_NONE, NULL, &error);
+  if (monitor == NULL)
+    {
+      g_autofree char *path = g_file_get_path (bwrapinfo);
+      g_error ("Failed to monitor %s: %s", path, error->message);
+      return;
+    }
+
+  monitor_data = g_new0 (BwrapinfoMonitorData, 1);
+  monitor_data->instance = g_steal_pointer (&instance);
+  monitor_data->pid = data->pid;
+  g_signal_connect (monitor, "changed", G_CALLBACK (bwrapinfo_changed), monitor_data);
+
+  check_child_pid_status (monitor, monitor_data);
+}
```

Comment:
This seems racy in general. There is no guarantee that we'll reach this code before the file is written, and in that case we would never get any change event. Also, file notifiers like this are super complicated internally and can run out of inotify handles and whatnot. I would recommend to just do the check, then if its no there yet, just do a loop via an timeout instead of the file monitor.

Reply:
Do note that in flatpak-run.c, the message is sent to the fd *after* bwrapinfo.json is created, so it should at least exist at this point. That being said if you prefer a timeout I can do that, would 100ms be fine? (I don't want to make it too large, since if you're spawning several processes it could cause visible delays.)

Reply:
Knowing that it exists is perhaps nice, but the problem is that it could be ready already and never see any change so will never trigger the read. 
I think what you want to do is immediately try to read the file, then do a say 1msec timeout and after that back it down to say 100msec.